### DISCUSSION TITLE
Can somebody explain @ObjectDependencies to me?

### DISCUSSION THREAD

**Discussion by bschmalhofer** [RotherOSS/otobo#1048] [Q&A] [answered]
Apparently I can't wrap my head around `@ObjectDependencies`. My understanding is that the dependencies declaration is used in `Kernel::System::ObjectsDiscard()` for making sure that the modules that depend on the to be discarded module are discarded first. That makes sense. But let's take a look at a concrete example. `Kernel::GenericInterface::Provider`declares that it depends on `Kernel::GenericInterface::Errorhandling`. So when K::G::E is discarded, than K::G::P is discarded too. But why should K::G::P care when K::G::E has been discarded? Whenever K::G::P needs a K::G::E, it gets the singleton that has been constructed with the current object params. I see no need to discard K::G::P. It would be a different when the state of K::G::P would depend on K::G::E, e.g. when K::G::E were used in the constructor of K::G::P for setting attributes.

**Comment by svenoe** [comment]
Mostly as I understand it, it is, that objects are discarded at all. Just deleting `$Self->{Objects}{K::G::P}` in the OM wouldn't have any effect on `$Self->{Objects}{K::G::E}`. The order is important only in special cases, have a look at `Kernel/System/Web/InterfaceAgent.pm` e.g., where `sub DESTROY` would instantiate a new `Kernel::System::Log`, which for that reason has to be deleted after K::S::W::I. (Ok, it's a bad example, as the K::S::W::I isn't managed by the OM atm, and also, if it is gone, the whole session will likely be gone, but just for the principle.)

**Reply by bschmalhofer** [reply]
Still confused. Objects being recreated in DESTROY is an issue. But that is a fairly special case. It would not warrant that every object that is used somewhere be declared as a dependency.

As I see it this leaves us with two reasons for declaring a dependency:
- The state of an object depends on the previous state of another object.
- The DESTROY() method depends on another object.

I'm not sure about the latter point, as object recreation should be fine. The point being, that the constructor uses the new params.

**Reply by svenoe** [reply]
Taking everything back - I didn't properly read it. It is the state you mentioned, as we discussed. E.g. `$Self->{ConfigObject}` in `Kernel/System/User/Preferences/DB.pm`.

**Comment by bschmalhofer** [comment]
I now have a better idea of the use of the object dependencies. So, I propose to adapt the code policy, See https://github.com/RotherOSS/CodePolicy/issues/63. Closing this discussion.

**Reply by bschmalhofer** [reply]
This is the answer. Can I mark it as such?

**Comment by bschmalhofer** [comment, marked as answer]
Answer was found.
